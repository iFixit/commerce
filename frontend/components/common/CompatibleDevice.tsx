import { Flex, Img, Text } from '@chakra-ui/react';
import type { Product } from '@pages/api/nextjs/cache/product';
import * as React from 'react';

export type CompatibleDeviceProps = {
   device: NonNullable<Product['compatibility']>['devices'][number];
   maxModelLines?: number;
};

export function CompatibleDevice({
   device,
   maxModelLines = 0,
}: CompatibleDeviceProps) {
   const variants = React.useMemo(
      () => device.variants.filter(Boolean),
      [device.variants]
   );
   const [visibleVariants, hiddenVariantCount] = React.useMemo(
      () =>
         maxModelLines > 0 && variants.length > maxModelLines
            ? [
                 variants.slice(0, maxModelLines - 1),
                 variants.length - maxModelLines + 1,
              ]
            : [variants, 0],
      [maxModelLines, variants]
   );
   return (
      <>
         <Img
            src={device.imageUrl}
            alt={device.deviceName ?? ''}
            w="12"
            mr="3"
            borderWidth="1px"
            borderStyle="solid"
            borderColor="gray.300"
            borderRadius="4px"
         />
         <Flex
            minH="12"
            flexDir="column"
            alignSelf="flex-start"
            justifyContent="center"
         >
            <Text
               my="2px"
               fontWeight="medium"
               _groupHover={{ color: 'brand.500' }}
            >
               {device.deviceName}
            </Text>
            <Flex flexDir="column">
               {visibleVariants.map((variant) => (
                  <Text
                     key={variant}
                     lineHeight="short"
                     fontSize="xs"
                     color="gray.600"
                  >
                     {variant}
                  </Text>
               ))}
            </Flex>
            {hiddenVariantCount > 0 && (
               <Text mb="2px" lineHeight="short" fontSize="xs" color="gray.600">
                  And {hiddenVariantCount} other model
                  {hiddenVariantCount > 1 ? 's' : ''}...
               </Text>
            )}
         </Flex>
      </>
   );
}
