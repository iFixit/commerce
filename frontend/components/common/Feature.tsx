import { Button, Flex, useToast, Text } from '@chakra-ui/react';
import { GetServerSidePropsContext } from 'next';
import Link from 'next/link';
import {
   ComponentType,
   PropsWithChildren,
   ReactNode,
   createContext,
   useCallback,
   useContext,
   useEffect,
} from 'react';
import {
   Cookies as ReactCookies,
   CookiesProvider,
   useCookies,
} from 'react-cookie';

export enum Features {
   ShowFeatureToasts = 'show-feature-toasts',
}

export function getAllFeatures(): Features[] {
   return Object.values(Features);
}

type Cookies = {
   [key: string]: string;
};

export type WithSSRFeatureCookies<T> = T & { cookies: Cookies };

export type FeatureContext = {
   features: Features[];
   isEnabled: (feature: Features) => boolean;
   setEnabled: (feature: Features, enabled: boolean) => void;
   toggleFeature: (feature: Features) => void;
};

const FeatureContext = createContext<FeatureContext | null>(null);

export function useFeatureContext(): FeatureContext {
   const context = useContext(FeatureContext);
   if (!context) {
      throw new Error('useFeatureContext must be used within a FeatureContext');
   }
   return context;
}

function warn(message: string, color = 'orange') {
   console.log('%c' + message, 'color:' + color);
}

export function withSSRFeatureCookies<P>(
   getServerSideProps: (
      context: GetServerSidePropsContext
   ) => Promise<{ props: P }>
): (
   context: GetServerSidePropsContext
) => Promise<{ props: WithSSRFeatureCookies<P> }> {
   return async (context: GetServerSidePropsContext) => {
      const props = await getServerSideProps(context);

      const cookies = Object.fromEntries(
         Object.entries(context.req.cookies || {}).filter(
            ([_key, value]) => value !== undefined
         )
      ) as Cookies;

      const enabledViaQueryParams = getAllFeatures().filter((feature) =>
         getQueryBoolean(context.query[feature.toString()])
      );

      enabledViaQueryParams.forEach((feature) => {
         warn(`Feature '${feature}' is was ENABLED via query param`);
         cookies[feature.toString()] = 'true';
         context.res.setHeader(
            'Set-Cookie',
            `${feature.toString()}=true; Path=/`
         );
      });

      return {
         props: {
            ...props.props,
            cookies: cookies,
         },
      };
   };
}

export function WithFeatureContextProvider<
   P extends WithSSRFeatureCookies<{ children?: ReactNode }>
>(Component: ComponentType<P>) {
   return function (props: P) {
      return (
         <FeatureProvider cookies={props.cookies}>
            <Component {...props} />
         </FeatureProvider>
      );
   };
}

export function WithFeature({
   enabled,
   disabled,
   children,
}: {
   enabled?: Features;
   disabled?: Features;
   children?: ReactNode;
}) {
   if (enabled && disabled) {
      throw new Error('WithFeature: only pass enable or disable, not both');
   }
   const { isEnabled } = useFeatureContext();

   const featureToCheck = enabled || disabled;

   if (!featureToCheck) {
      return <>{children}</>;
   }

   const wantsFeature = isEnabled(featureToCheck);

   if (enabled && !wantsFeature) {
      return null;
   }
   if (disabled && wantsFeature) {
      return null;
   }

   return <>{children}</>;
}

type FeatureProviderProps = WithSSRFeatureCookies<{
   children?: ReactNode;
}>;

export const FeatureProvider = withCookieProvider<FeatureProviderProps>(
   function FeatureContextProvider({
      children,
   }: {
      children?: React.ReactNode;
   }) {
      const [cookies, setCookie, removeCookie] = useCookies();
      const features = getAllFeatures();

      const isEnabled = useCallback(
         (feature: Features) => {
            return cookies[feature.toString()] === 'true';
         },
         [cookies]
      );

      const setEnabled = useCallback(
         (feature: Features, enabled: boolean) => {
            if (!enabled) {
               removeCookie(feature.toString(), {
                  path: '/',
               });
               return;
            }

            setCookie(feature.toString(), enabled.toString(), {
               path: '/',
            });
         },
         [setCookie, removeCookie]
      );

      const toggleFeature = useCallback(
         (feature: Features) => {
            setEnabled(feature, !isEnabled(feature));
         },
         [setEnabled, isEnabled]
      );

      features.forEach((feature) => {
         const enabled = isEnabled(feature);
         warn(`Feature '${feature}' is ${enabled ? 'ENABLED' : 'DISABLED'}`);
      });

      const toast = useToast();

      useEffect(() => {
         if (!isEnabled(Features.ShowFeatureToasts)) {
            return;
         }

         const enabledFeatures = features.filter((feature) =>
            isEnabled(feature)
         );

         const disabledFeatures = features.filter(
            (feature) => !isEnabled(feature)
         );

         if (enabledFeatures.length || disabledFeatures.length) {
            toast.closeAll();
         }

         if (enabledFeatures.length) {
            toast({
               title: 'Enabled Features',
               description: (
                  <FeatureToastContent
                     features={enabledFeatures}
                     colorScheme="green"
                  />
               ),
               status: 'success',
               duration: 5000,
               isClosable: true,
            });
         }

         if (disabledFeatures.length) {
            toast({
               title: 'Disabled Features',
               description: (
                  <FeatureToastContent
                     features={disabledFeatures}
                     colorScheme="orange"
                  />
               ),
               status: 'warning',
               duration: 5000,
               isClosable: true,
            });
         }
      }, [features, isEnabled, toast]);
      const context = {
         features,
         isEnabled,
         setEnabled,
         toggleFeature,
      };

      return (
         <FeatureContext.Provider value={context}>
            {children}
         </FeatureContext.Provider>
      );
   }
);

function FeatureToastContent({
   features,
   colorScheme,
}: {
   features: Features[];
   colorScheme: string;
}) {
   return (
      <Flex flexDirection="column">
         {features.map((feature) => (
            <Text key={feature}>{feature}</Text>
         ))}
         <Button
            as={Link}
            size="sm"
            href="/dev/features"
            variant="solid"
            colorScheme={colorScheme}
         >
            See Features Page
         </Button>
      </Flex>
   );
}

function withCookieProvider<
   P extends {
      children?: ReactNode;
      cookies: {
         [key: string]: string;
      };
   }
>(Component: ComponentType<PropsWithChildren<P>>) {
   const isBrowser = typeof window !== 'undefined';
   return function (props: P) {
      return (
         <CookiesProvider
            cookies={isBrowser ? undefined : new ReactCookies(props.cookies)}
         >
            <Component {...props} />
         </CookiesProvider>
      );
   };
}

function getQueryBoolean(
   query: string | string[] | undefined | null,
   options: { considerEmptyAsTrue?: boolean } = { considerEmptyAsTrue: true }
): boolean {
   if (query === undefined || query === null) {
      return false;
   }

   if (Array.isArray(query)) {
      return query.some((q) => getQueryBoolean(q, options));
   }

   const boolValue = /true|0/i.test(query);

   return options.considerEmptyAsTrue ? query === '' : boolValue;
}
