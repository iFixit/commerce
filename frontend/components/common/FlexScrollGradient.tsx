import {
   Flex,
   FlexProps,
   SystemStyleObject,
   useMergeRefs,
} from '@chakra-ui/react';
import { useRef, forwardRef, useState } from 'react';
import { useIsomorphicLayoutEffect } from '@ifixit/ui';

type InnerFlexStyling = {
   _after: SystemStyleObject;
   _before: SystemStyleObject;
   overflowX?: FlexProps['overflowX'];
   overflowY?: FlexProps['overflowY'];
};

const gradientSizes = { base: 50, sm: 75, md: 90, lg: 100 };

const sharedStyle: SystemStyleObject = {
   content: '""',
   position: 'absolute',
   height: '100%',
   pointerEvents: 'none',
   zIndex: 9999,
};

function easeInOutQuad(x: number): number {
   return x < 0.5 ? 2 * x * x : 1 - Math.pow(-2 * x + 2, 2) / 2;
}

enum FlexScrollDirection {
   ROW,
   COLUMN,
}

function getScrollDirection(el: HTMLElement): FlexScrollDirection {
   const { flexDirection } = window.getComputedStyle(el);
   if (flexDirection === 'column') {
      return FlexScrollDirection.COLUMN;
   } else {
      return FlexScrollDirection.ROW;
   }
}

function getGradientsBaseStyle(
   scrollDirection: FlexScrollDirection,
   gradientSizePX: number,
   color = '#f9fafb'
) {
   const beforeStart =
      scrollDirection === FlexScrollDirection.ROW ? 'left' : 'top';
   const afterStart =
      scrollDirection === FlexScrollDirection.ROW ? 'right' : 'bottom';

   const beforeBackgroundDegrees =
      scrollDirection === FlexScrollDirection.ROW ? 270 : 0;
   const afterBackgroundDegrees =
      scrollDirection === FlexScrollDirection.ROW ? 90 : 180;

   const gradientSizeProp =
      scrollDirection === FlexScrollDirection.ROW ? 'width' : 'height';

   return {
      before: {
         ...sharedStyle,
         [gradientSizeProp]: `${gradientSizePX}px`,
         [beforeStart]: '0px',
         background: `linear-gradient(${beforeBackgroundDegrees}deg, rgba(249, 250, 251, 0) 0%, ${color} 106.41%)`,
      },
      after: {
         ...sharedStyle,
         [gradientSizeProp]: `${gradientSizePX}px`,
         [afterStart]: '0px',
         background: `linear-gradient(${afterBackgroundDegrees}deg, rgba(249, 250, 251, 0) 0%, ${color} 106.41%)`,
      },
   };
}

function getGradientStyleProps(el: HTMLElement): InnerFlexStyling {
   const flexScrollDirection = getScrollDirection(el);
   const {
      // Used for both calcs
      clientWidth,
      clientHeight,
      // Used for ROWs
      scrollWidth,
      scrollLeft,
      // Used for COLUMNs
      scrollHeight,
      scrollTop,
   } = el;

   const isScrollable =
      flexScrollDirection === FlexScrollDirection.ROW
         ? scrollWidth > clientWidth
         : scrollHeight > clientHeight;

   if (!isScrollable) {
      return {
         _before: { opacity: 0, height: 0 },
         _after: { opacity: 0, height: 0 },
      };
   }

   const start =
      flexScrollDirection === FlexScrollDirection.ROW ? scrollLeft : scrollTop;
   const windowSpace =
      flexScrollDirection === FlexScrollDirection.ROW
         ? clientWidth
         : clientHeight;
   const gradientSpace =
      flexScrollDirection === FlexScrollDirection.ROW
         ? clientHeight
         : clientWidth;
   const scrollSpace =
      flexScrollDirection === FlexScrollDirection.ROW
         ? scrollWidth
         : scrollHeight;
   const sizeProp =
      flexScrollDirection === FlexScrollDirection.ROW ? 'height' : 'width';

   const scrollPosition = start + windowSpace;
   const gradientSizePX = getGradientSize(windowSpace);
   const afterStart = scrollSpace - gradientSizePX;

   const size = `${gradientSpace}px`;

   const beforeGradientShownPercent =
      start < gradientSizePX ? start / gradientSizePX : 1;

   const afterGradientShownPercent =
      scrollPosition >= afterStart
         ? 1 - (scrollPosition - afterStart) / gradientSizePX
         : 1;

   const beforeGradientOpacity = easeInOutQuad(
      Math.max(0, beforeGradientShownPercent)
   );

   const afterGradientOpacity = easeInOutQuad(
      Math.max(0, afterGradientShownPercent)
   );

   const { before, after } = getGradientsBaseStyle(
      flexScrollDirection,
      gradientSizePX
   );

   const { borderRadius } = window.getComputedStyle(el);

   const overflowX =
      flexScrollDirection === FlexScrollDirection.ROW ? 'auto' : undefined;
   const overflowY =
      flexScrollDirection === FlexScrollDirection.ROW ? undefined : 'auto';

   return {
      _before: {
         ...before,
         opacity: beforeGradientOpacity,
         [sizeProp]: size,
         borderRadius,
      },
      _after: {
         ...after,
         opacity: afterGradientOpacity,
         [sizeProp]: size,
         borderRadius,
      },
      overflowX,
      overflowY,
   };
}

function getGradientSize(elSize: number): number {
   if (elSize < 576) {
      return gradientSizes.base;
   } else if (elSize < 768) {
      return gradientSizes.sm;
   } else if (elSize < 1028) {
      return gradientSizes.md;
   } else {
      return gradientSizes.lg;
   }
}

export const FlexScrollGradient = forwardRef(function FlexScrollGradient(
   {
      nestedFlexProps,
      ...props
   }: FlexProps & {
      nestedFlexProps?: FlexProps;
   },
   ref
) {
   const internalRef = useRef<HTMLDivElement>();
   const refs = useMergeRefs(internalRef, ref);

   const [flexScrollState, setFlexScrollState] = useState<InnerFlexStyling>({
      _before: { opacity: 0 },
      _after: { opacity: 0 },
   });

   useIsomorphicLayoutEffect(() => {
      const el = internalRef.current;
      if (!el) {
         return;
      }

      const doMeasure = (el: HTMLElement) => {
         const state = getGradientStyleProps(el);
         setFlexScrollState(() => state);
      };

      const measure = () => {
         doMeasure(el);
      };

      measure();

      el.addEventListener('transitionend', measure, { passive: true });
      el.addEventListener('scroll', measure, { passive: true });
      window.addEventListener('resize', measure, { passive: true });

      return () => {
         el.removeEventListener('transitionend', measure);
         el.removeEventListener('scroll', measure);
         window.removeEventListener('resize', measure);
      };
   }, [internalRef]);

   if (props.position) {
      throw new Error(
         'FlexScrollGradient: position prop is not allowed. To get the gradient working we must use position relative. Wrap your component in another layer to handle positioning.'
      );
   }

   return (
      <Flex {...props} position="relative" isolation="isolate">
         <Flex
            ref={refs}
            {...nestedFlexProps}
            overflowX={flexScrollState.overflowX}
            overflowY={flexScrollState.overflowY}
            _before={flexScrollState._before}
            _after={flexScrollState._after}
         >
            {props.children}
         </Flex>
      </Flex>
   );
});
