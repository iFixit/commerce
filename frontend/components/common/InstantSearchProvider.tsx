import { useSafeLayoutEffect } from '@chakra-ui/react';
import { useSearchCache } from '@components/product-list/sections/FilterableProductsSection/useSearchCache';
import { ALGOLIA_APP_ID } from '@config/env';
import {
   decodeDeviceItemType,
   encodeDeviceItemType,
} from '@helpers/product-list-helpers';
import { cypressWindowLog } from '@helpers/test-helpers';
import { useAuthenticatedUser } from '@ifixit/auth-sdk';
import { usePrevious } from '@ifixit/ui';
import algoliasearch from 'algoliasearch/lite';
import { history } from 'instantsearch.js/es/lib/routers';
import { RouterProps } from 'instantsearch.js/es/middlewares';
import { UiState } from 'instantsearch.js/es/types';
import { mapValues } from 'lodash';
import { useRouter } from 'next/router';
import * as React from 'react';
import {
   InstantSearch,
   InstantSearchServerState,
   InstantSearchSSRProvider,
} from 'react-instantsearch-hooks-web';

type InstantSearchProviderProps = React.PropsWithChildren<AlgoliaProps>;

export type AlgoliaProps = {
   url: string;
   indexName: string;
   serverState?: Partial<InstantSearchServerState>;
   apiKey: string;
};

type RouteState = Partial<{
   q: string;
   p: number;
   filter: Record<string, any>;
   // range: Record<string, string>;
}>;

type IndexUiState = Record<string, any>;

export function InstantSearchProvider({
   children,
   url,
   indexName,
   serverState,
   apiKey,
}: InstantSearchProviderProps) {
   const user = useAuthenticatedUser();
   cypressWindowLog({ userLoaded: user.isFetched });
   const algoliaApiKey = user.data?.algoliaApiKeyProducts || apiKey;
   const previousApiKey = usePrevious(algoliaApiKey);

   const algoliaClient = React.useMemo(() => {
      return algoliasearch(ALGOLIA_APP_ID, algoliaApiKey);
   }, [algoliaApiKey]);

   const router = useRouter();

   // Currently, Algolia routing does not play well with Next.js routing, since Next.js
   // is not aware of url changes that happens without interacting with its builtin router.
   // To make popstate work (i.e. whenever the browser back button is pressed), we do a
   // a full page reload as a workaround.
   // This should be only a temporary workaround until Algolia routing is fixed.
   useSafeLayoutEffect(() => {
      const handleRouteChange = () => {
         // When the back button is pressed, the popstate event is fired, and
         // Algolia tries to re-render the page (although in a broken state).
         // To avoid seeing this broken re-render we hide the page while we wait for
         // the page to be reloaded.
         window.document.body.hidden = true;
         window.location.reload();
      };
      window.addEventListener('popstate', handleRouteChange);
      return () => {
         window.removeEventListener('popstate', handleRouteChange);
      };
   }, [router]);

   // We're using this to make `InstantSearch` unmount at every re-render, since
   // as of version 6.28.0 it breaks when it re-renders. Re-rendering though
   // should be relatively infrequent in this component, so this should be fine.
   const renderCount = useCountRenders();

   const routing: RouterProps<UiState, RouteState> = {
      stateMapping: {
         stateToRoute(uiState) {
            const indexUiState = uiState[indexName];
            const routeState: RouteState = {};
            if (indexUiState.query) {
               routeState.q = indexUiState.query;
            }
            if (indexUiState.page) {
               routeState.p = indexUiState.page;
            }
            if (indexUiState.refinementList) {
               routeState.filter = indexUiState.refinementList;
            }
            // if (indexUiState.range != null) {
            //    routeState.range = indexUiState.range;
            // }
            return routeState;
         },
         routeToState(routeState: RouteState) {
            const stateObject: IndexUiState = {};
            if (routeState.q != null) {
               stateObject.query = routeState.q;
            }
            if (routeState.p != null) {
               stateObject.page = routeState.p;
            }
            if (routeState.filter != null) {
               stateObject.refinementList = routeState.filter;
            }
            // if (routeState.range != null) {
            //    stateObject.range = routeState.range;
            // }
            return {
               [indexName]: stateObject,
            };
         },
      },
      router: history({
         getLocation() {
            if (typeof window === 'undefined') {
               return new URL(url) as unknown as Location;
            }

            return window.location;
         },
         createURL({ qsModule, routeState, location }) {
            const baseUrl = getBaseOrigin(location);
            const pathParts = location.pathname
               .split('/')
               .filter((part) => part !== '');
            const partsOrTools = pathParts.length >= 1 ? pathParts[0] : '';
            const deviceHandle = pathParts.length >= 2 ? pathParts[1] : '';
            const itemTypeHandle = pathParts.length >= 3 ? pathParts[2] : '';

            let path = '';
            if (partsOrTools) {
               path += `/${partsOrTools}`;
               if (deviceHandle) {
                  path += `/${deviceHandle}`;
                  const raw: string | string[] | undefined =
                     routeState.filter?.['facet_tags.Item Type'];
                  const itemType = Array.isArray(raw) ? raw[0] : raw;
                  if (itemType?.length) {
                     const encodedItemType = encodeDeviceItemType(itemType);
                     path += `/${encodedItemType}`;
                  } else if (renderCount === 1 && itemTypeHandle) {
                     // Prevents a bug when visiting /Parts/iPhone/Cables
                     // that it would redirect to /Parts/iPhone
                     path += `/${itemTypeHandle}`;
                  }
               }
            }

            const filterCopy = { ...routeState.filter };
            if (partsOrTools && deviceHandle) {
               // Item Type is the slug on device pages, not in the query.
               delete filterCopy['facet_tags.Item Type'];
            }
            const queryString = qsModule.stringify(
               { ...routeState, filter: filterCopy },
               {
                  addQueryPrefix: true,
                  arrayFormat: 'indices',
               }
            );

            return `${baseUrl}${path}${queryString}`;
         },
         parseURL({ qsModule, location }) {
            const pathParts = location.pathname
               .split('/')
               .filter((part) => part !== '');
            const deviceHandle = pathParts.length >= 2 ? pathParts[1] : '';
            const itemType = pathParts.length >= 3 ? pathParts[2] : '';

            const {
               q = '',
               p,
               filter = {},
               // range = {},
            } = qsModule.parse(location.search.slice(1));

            const decodedFilters = decodeParsedQuery(filter);
            if (deviceHandle && itemType) {
               decodedFilters['facet_tags.Item Type'] = [
                  decodeDeviceItemType(decodeURIComponent(itemType)),
               ];
            }

            return {
               q: decodeURIComponent(String(q)),
               p: Number(p),
               filter: decodedFilters,
               // range: decodeParsedQuery(range),
            };
         },
      }),
   };

   return (
      <InstantSearchSSRProvider {...serverState}>
         <InstantSearch
            key={renderCount}
            searchClient={algoliaClient}
            indexName={indexName}
            routing={routing}
         >
            <RefreshSearchResults
               apiKey={algoliaApiKey}
               prevApiKey={previousApiKey}
            />
            {children}
         </InstantSearch>
      </InstantSearchSSRProvider>
   );
}

function useCountRenders() {
   const countRef = React.useRef(0);
   countRef.current++;
   return countRef.current;
}

function decodeParsedQuery(parsed: any) {
   return typeof parsed === 'object'
      ? mapValues<Record<string, any>, any>(parsed, (parsedValues) => {
           return Array.isArray(parsedValues)
              ? parsedValues.map((v: any) =>
                   typeof v === 'string' ? decodeURIComponent(v) : v
                )
              : decodeURIComponent(parsedValues);
        })
      : {};
}

type RefreshSearchResultsProps = {
   apiKey: string;
   prevApiKey: string | undefined;
};

/**
 * Refreshes search results if the API key changes
 * (but only refreshes once for each api key change).
 */
function RefreshSearchResults({
   apiKey,
   prevApiKey,
}: RefreshSearchResultsProps): null {
   const { refresh } = useSearchCache();
   useSafeLayoutEffect(() => {
      const isFirstRender = !prevApiKey;
      const hasApiKeyChanged = prevApiKey !== apiKey;
      if (hasApiKeyChanged && !isFirstRender) {
         refresh();
      }
   }, [apiKey, prevApiKey]);
   return null;
}

function getBaseOrigin(location: Location): string {
   if (typeof window === 'undefined' && process.env.NEXT_PUBLIC_IFIXIT_ORIGIN) {
      // On the server, use the IFIXIT_ORIGIN url
      // This ensures that the SSR produces the correct links on Vercel
      // (where the Host header doesn't match the page URL.)
      const publicOrigin = new URL(process.env.NEXT_PUBLIC_IFIXIT_ORIGIN);
      return publicOrigin.origin;
   }
   return location.origin;
}
