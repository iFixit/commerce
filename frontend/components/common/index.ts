export * from './AppProviders';
export * from './ProductCard';
export * from './SecondaryNavbar';
export * from './PageBreadcrumb';
export * from './useSearchCache';
export * from './ProductRating';
export * from './CompatibleDevice';
export * from './PrerenderedHTML';
