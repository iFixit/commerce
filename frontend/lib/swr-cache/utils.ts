import isPlainObject from 'lodash/isPlainObject';
import { z } from 'zod';

export interface CacheAdapter {
   name?: string;
   set(key: string, value: CacheEntry): Promise<void>;
   get(key: string): Promise<CacheEntry | null>;
   delete(key: string): Promise<void>;
}

const CacheEntrySchema = z.object({
   metadata: z.object({
      ttl: z.number().positive().optional(),
      staleWhileRevalidate: z.number().optional(),
      createdAt: z.number(),
   }),
   value: z.unknown(),
});

type CacheEntry = z.infer<typeof CacheEntrySchema>;

interface CreateCacheEntryOptions {
   ttl?: number;
   staleWhileRevalidate?: number;
}

export const createCacheEntry = <T = unknown>(
   value: T,
   options: CreateCacheEntryOptions
): CacheEntry => {
   const { ttl, staleWhileRevalidate } = options;
   return {
      metadata: {
         ttl,
         staleWhileRevalidate: ensurePositiveNumber(staleWhileRevalidate),
         createdAt: Date.now(),
      },
      value,
   };
};

export const isValidCacheEntry = (entry: unknown): entry is CacheEntry => {
   const validation = CacheEntrySchema.safeParse(entry);
   if (!validation.success) {
      return false;
   }
   return isFresh(validation.data) || isStale(validation.data);
};

export const isStale = (entry: CacheEntry) => {
   const now = Date.now();
   const {
      ttl = Infinity,
      staleWhileRevalidate = 0,
      createdAt,
   } = entry.metadata;

   return !isFresh(entry) && createdAt + ttl + staleWhileRevalidate > now;
};

export const createCacheKey = (keyPrefix: string, variables: unknown) => {
   return `${keyPrefix}:${stableHash(variables)}`;
};

export const sleep = (ms: number) =>
   new Promise((resolve) => setTimeout(resolve, ms));

export const ensurePositiveNumber = (value: number | null | undefined) => {
   return value == null ? 0 : Math.max(0, value);
};

const isFresh = (entry: CacheEntry) => {
   const now = Date.now();
   const { ttl = Infinity, createdAt } = entry.metadata;
   return createdAt + ttl > now;
};

const stableHash = (variables: unknown): string => {
   return JSON.stringify(variables, (_, val) =>
      isPlainObject(val)
         ? Object.keys(val)
              .sort()
              .reduce((result, key) => {
                 result[key] = val[key];
                 return result;
              }, {} as any)
         : val
   );
};
