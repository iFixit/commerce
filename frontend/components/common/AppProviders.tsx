import { ChakraProvider, extendTheme } from '@chakra-ui/react';
import { IFIXIT_ORIGIN } from '@config/env';
import { AppProvider } from '@ifixit/app';
import { theme } from '@ifixit/ui';
import * as React from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { AlgoliaProps, InstantSearchProvider } from './InstantSearchProvider';

const customTheme = extendTheme({
   ...theme,
   sizes: {
      ...theme.sizes,
      header: '68px',
   },
   zIndices: {
      ...theme.zIndices,
      header: 2000,
   },
   styles: {
      ...theme.styles,
      global: {
         ...theme.styles?.global,
         body: {
            // @ts-ignore
            ...theme.styles?.global?.body,
            touchAction: 'pan-x pan-y',
         },
      },
   },
});

const queryClient = new QueryClient();

export type WithProvidersProps<T> = T & { appProps: AppProvidersProps };

export type AppProvidersProps = {
   algolia?: AlgoliaProps;
};

export function AppProviders({
   children,
   algolia,
}: React.PropsWithChildren<AppProvidersProps>) {
   const markup = (
      <ChakraProvider theme={customTheme}>{children}</ChakraProvider>
   );

   return (
      <AppProvider ifixitOrigin={IFIXIT_ORIGIN}>
         <QueryClientProvider client={queryClient}>
            {algolia ? (
               <InstantSearchProvider {...algolia}>
                  {markup}
               </InstantSearchProvider>
            ) : (
               markup
            )}
         </QueryClientProvider>
      </AppProvider>
   );
}
