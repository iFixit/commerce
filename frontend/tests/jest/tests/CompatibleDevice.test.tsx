import { CompatibleDevice } from '@components/common';
import { renderWithAppContext } from '../utils';
import { getMockProduct } from '../utils';

describe('CompatibleDevice', () => {
   it('renders and matches the snapshot', () => {
      let product = getMockProduct();
      let device = product!.compatibility!.devices[0];

      // @ts-ignore
      const { getByText, asFragment } = renderWithAppContext(
         <CompatibleDevice device={device} />
      );

      // getByText throws an error if it can't find the text
      (expect(() => getByText(/other models.../i)) as any).toThrow();
      (expect(asFragment()) as any).toMatchSnapshot();
   });

   it('renders and matches the snapshot with truncated variants', async () => {
      let product = getMockProduct();
      let device = product!.compatibility!.devices[0];

      // @ts-ignore
      const { getByText, asFragment } = renderWithAppContext(
         <CompatibleDevice device={device} truncate={1} />
      );

      (expect(getByText(/other models.../i)) as any).toBeTruthy();
      (expect(asFragment()) as any).toMatchSnapshot();
   });
});
