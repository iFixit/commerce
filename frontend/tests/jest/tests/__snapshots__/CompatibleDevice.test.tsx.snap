// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompatibleDevice renders and matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1tbib0t"
  >
    <a
      class="chakra-link css-spn4bz"
      href="/Device/iPhone_6s_Plus"
    >
      <img
        alt="iPhone 6s Plus"
        class="chakra-image css-ky1liq"
        src="https://mmcelvain.cominor.com/igi/cMVbyIbIrTEbi2j5.thumbnail"
      />
    </a>
    <div
      class="css-qi8j5b"
    >
      <a
        class="css-ce03x1"
        href="/Device/iPhone_6s_Plus"
      >
        iPhone 6s Plus
      </a>
      <div
        class="css-j7qwjs"
      >
        <p
          class="chakra-text css-4137gr"
        >
          A1634 US AT&T Locked or SIM Free
        </p>
        <p
          class="chakra-text css-4137gr"
        >
          A1687 US Sprint/Verizon and Global
        </p>
        <p
          class="chakra-text css-4137gr"
        >
          A1699 Mainland China
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`CompatibleDevice renders and matches the snapshot with truncated variants 1`] = `
<DocumentFragment>
  <div
    class="css-1tbib0t"
  >
    <a
      class="chakra-link css-spn4bz"
      href="/Device/iPhone_6s_Plus"
    >
      <img
        alt="iPhone 6s Plus"
        class="chakra-image css-ky1liq"
        src="https://mmcelvain.cominor.com/igi/cMVbyIbIrTEbi2j5.thumbnail"
      />
    </a>
    <div
      class="css-qi8j5b"
    >
      <a
        class="css-ce03x1"
        href="/Device/iPhone_6s_Plus"
      >
        iPhone 6s Plus
      </a>
      <div
        class="css-j7qwjs"
      >
        <p
          class="chakra-text css-4137gr"
        >
          A1634 US AT&T Locked or SIM Free
        </p>
      </div>
      <p
        class="chakra-text css-1ji9qma"
      >
        And 2 more...
      </p>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
